src/libopenrave-core/generictrajectory.cpp

```diff
@@ -319,26 +409,76 @@ class GenericTrajectory : public TrajectoryBase
+        // NOTE: Ignore 'options' argument for now
+
+        // Write binary file header
+        WriteBinaryUInt16(O, MAGIC_NUMBER);
+        WriteBinaryUInt16(O, VERSION_NUMBER);
+
+        /* Store meta-data */
+
+        // Indicate size of meta data
+        const ConfigurationSpecification& spec = this->GetConfigurationSpecification();
+        const uint16_t numGroups = spec._vgroups.size();
+        WriteBinaryUInt16(O, numGroups);
+
+        FOREACHC(itgroup, spec._vgroups)
+        {
+            WriteBinaryString(O, itgroup->name);   // Writes group name
+            WriteBinaryInt(O, itgroup->offset);    // Writes offset
+            WriteBinaryInt(O, itgroup->dof);       // Writes dof
+            WriteBinaryString(O, itgroup->interpolation);  // Writes interpolation
         }
-        if( GetReadableInterfaces().size() > 0 ) {
-            xmlreaders::StreamXMLWriterPtr writer(new xmlreaders::StreamXMLWriter("readable"));
-            FOREACHC(it, GetReadableInterfaces()) {
-                BaseXMLWriterPtr newwriter = writer->AddChild(it->first);
-                it->second->Serialize(newwriter,options);
+
+        /* Store data waypoints */
+        WriteBinaryVector(O, this->_vtrajdata);
+    }
+
+    InterfaceBasePtr deserialize(std::istream& I)
+    {
+        // Check whether binary or XML file
+        stringstream::streampos pos = I.tellg();  // Save old position
+        uint16_t binaryFileHeader = 0;
+        ReadBinaryUInt16(I, binaryFileHeader);
+
+        // Read binary trajectory files
+        if (binaryFileHeader == MAGIC_NUMBER)
+        {   
+            uint16_t versionNumber = 0;
+            ReadBinaryUInt16(I, versionNumber);
+
+            if (versionNumber != VERSION_NUMBER)
+            {
+                throw OPENRAVE_EXCEPTION_FORMAT(_("unsupported trajectory format version %d "),versionNumber,ORE_InvalidArguments);
+            }
+                
+            /* Read metadata */
+
+            // Read number of groups
+            uint16_t numGroups = 0;
+            ReadBinaryUInt16(I, numGroups);
+
+            ConfigurationSpecification spec;
+            spec._vgroups.resize(numGroups);
+            FOREACH(itgroup, spec._vgroups)
+            {
+                ReadBinaryString(I, itgroup->name);             // Read group name
+                ReadBinaryInt(I, itgroup->offset);              // Read offset
+                ReadBinaryInt(I, itgroup->dof);                 // Read dof
+                ReadBinaryString(I, itgroup->interpolation);    // Read interpolation
             }
-            writer->Serialize(O);
+            this->Init(spec);
+
+            /* Read trajectory data */
+            ReadBinaryVector(I, this->_vtrajdata);
+            return shared_from_this();
         }
-        O << "</trajectory>" << endl;
+        
+        // Call on parent class deserialize method for backwards compatible with old XML trajectory files
+        return TrajectoryBase::deserialize(I);
```

Comment:
@ShaysRebellion did you try deserializing with XML to see if it works. don't you need to call `seekg` to reset `I`?